python/taichi/lang/struct.py

```diff
@@ -49,6 +56,14 @@ def __init__(self, *args, **kwargs):
             raise TaichiSyntaxError(
                 "Custom structs need to be initialized using either dictionary or keyword arguments"
             )
+        self.methods = self.entries.pop("__struct_methods", {})
+        for name, method in self.methods.items():
+            # use MethodType to pass self (this object) to the method
+            setattr(self, name, MethodType(method, self))
+            if getattr(method, '_is_taichi_function', False) is True:
```

Comment:
I'm a little curious about why we need this. Doesn't `@ti.func` work?

BTW are we supposed to get these struct methods run in both Python scope and Taichi scope?

Reply:
So the fact that we're re-binding the function to a new object with the func() call seems to lose the ti.func bits and gives a warning.  

Python and taichi scope functions should work fine with this change.